### ISSUE TITLE
socket hang up

### ISSUE THREAD

**Issue by aaronovz1** [solana-labs/solana-web3.js#1426]
## Overview

We keep getting these occasional `socket hang up` errors which seems to cause our backend job to hang.

## Steps to reproduce

This is difficult to reproduce but the job which triggers it often is pretty heavy on async calls with many RPC requests firing off in parallel using the same `Connection` object.

## Description of bug

```
FetchError: request to https://rpcpool.com/[REDACTED] failed, reason: socket hang up
 at ClientRequest.<anonymous> (/usr/app/node_modules/node-fetch/lib/index.js:1491:11)
 at ClientRequest.emit (node:events:513:28)
 at TLSSocket.socketOnEnd (node:_http_client:526:9)
 at TLSSocket.emit (node:events:525:35)
```

**Comment by aaronovz1** [comment]
Possibly issue with outdated node-fetch as I'm seeing these errors as well after updating to 1.78
https://github.com/node-fetch/node-fetch/issues/1219

**Comment by steveluscher** [comment]
Possibly a duplicate of?

1. https://github.com/solana-labs/solana-web3.js/issues/1255#issuecomment-1563393009
2. https://github.com/solana-labs/solana/pull/29130

**Comment by steveluscher** [comment]
Try configuring your _own_ keep-alive timeout (with advice from your RPC provider) as mentioned [here](https://github.com/solana-labs/solana/pull/29130#issuecomment-1358065043), and report back!

**Comment by aaronovz1** [comment]
Thanks @steveluscher - chatting to Triton now.

What do you think about the these issues causing code to hang though? Smells like a secondary issue with error handling somewhere in either this package or node-fetch.

**Comment by steveluscher** [comment]
I drew diagrams over at #29130!

**Comment by steveluscher** [comment]
Unless your question is ‘why isn't the infra resilient, even in the face of this connection tracking bug’ to which my answer is: if you can't trust the connection tracking code, all bets are basically off.

**Comment by aaronovz1** [comment]
> Unless your question is ‘why isn't the infra resilient, even in the face of this connection tracking bug’ to which my answer is: if you can't trust the connection tracking code, all bets are basically off.

No, I'm asking why does the Solana web3 `Connection` seemingly cause the code path using it to hang indefinitely. The issue we're having is we first see the socket hang up issue, then shortly after the process/job hangs forever. I don't see how your diagram explains that - looks more like it should timeout and throw some kind of error, not hang.

**Comment by steveluscher** [comment]
It's been 6 months since I've had all the context loaded into my head, but here's what I remember:

* The client can only rely on what the proxy tells it. The proxy (load balancer in Triton's case) says the socket is open.
* What the proxy doesn't know is that the RPC has since gone away. The socket is, in fact, not healthy.
* The client can't tell the difference between there being a network error because of a transient issue, or there being a network error because the proxy and the RPC broke up, so it just keeps retrying according to the retry logic. This to you manifests as an infinite retry loop (ie. a hang).

If we start _guessing_ why the network is emitting errors, we will be wrong some amount of the time. The solution is either for:

1. The HTTP Agent on the client, the load balancer, and the RPC server to all agree on a keep-alive timeout
2. Set `httpAgent` to `false` on the client
